### PR TITLE
GLTFLoader: Deduplicate node names.

### LINKED ISSUE #15087: GLTFLoader: De-duplicate node names.
Names of nodes are user-specified, and not guaranteed to be unique within a glTF file. But because THREE.PropertyBinding relies on track names to target animation, having two nodes with the same name will cause animation to affect the wrong part of a model. We should de-duplicate node names on import, to ensure that every node ends up with a unique name. 

Filing to follow up on https://discourse.threejs.org/t/gltf-animation-bug-or-wrong-usage/4536/2.

### PR DESCRIPTION
**Pull request by donmccurdy**
Fixes #15087.

/cc @takahirox

### CODE REVIEW

**Comment by takahirox** [comment]
Good work. Probably renaming is reasonable to resolve the animation problem.

But minor concern. Polluting user or asset-author defined data (name) might break a certain workflow. For example, if a user has a program out of Three.js processing something only for nodes which have a certain name. Imagine one import gltf to Three.js, does something, and then exporting back to gltf. If the loader rename the program might not work correctly for exported gltf.

I think it may be kind of rare use case. It'd be ok to merge with this PR now and resolve my concern later.

Potential solutions may be

1. With #16412 we store original name. GLTFExporter restores the name when exporting.

2. Console warning if renamed. User can notice the rename. Cons is it can be tons of warning if many nodes are renamed.

3. Add rename option to the loader. Default is false. User needs to explicitly to turn it on for rename so one can notice the rename.

I speculate 1 sounds good?

**Comment by takahirox** [comment]
BTW, I'd think better to improve the algorithm.

```javascript
GLTFParser.prototype.assignUniqueName = function ( object, originalName ) {

	var name = originalName;

	for ( var i = 1; this.nodeNamesUsed[ name ]; ++ i ) {

		name = originalName + '_' + i;

	}

	this.nodeNamesUsed[ name ] = true;

	object.name = name;

};
```

The complexity is O(n) where n is the number of nodes in a gltf having the same name (for example empty name). One renaming spends O(n), so it spends O(n^2) for parsing entire gltf because renaming n times.

This is similar performance problem to the geometry cache performance problem we had and fixed. Hopefully O(1) algorithm is good.

File: examples/js/loaders/GLTFLoader.js [hunk 1]
```diff
@@ -2432,6 +2435,22 @@ THREE.GLTFLoader = ( function () {
...
+			name = originalName + '_' + i;
+
+		}
+
+		this.nodeNamesUsed[ name ] = true;
+
+		object.name = name;
+
+	};
+
```
Comment (takahirox) [inline examples/js/loaders/GLTFLoader.js]: Nit, and no strong preference. But what do you think of removing `object` argument and returning the unique name instead? The methods will have less dependency with `object`.

```javascript
GLTFParser.prototype.createUniqueName = function ( originalName ) {

    var name = originalName;

    for ( var i = 1; this.nodeNamesUsed[ name ]; ++ i ) {

        name = originalName + '_' + i;

    }

    this.nodeNamesUsed[ name ] = true;

    return name;

};

node.name = parser.createUniqueName( node.name );
```

**Comment by donmccurdy** [comment]
I’d prefer not to preserve the original name in extras, not to give a warning, and not to add a new option – this all seems too complex for a case that has never given us any reason to add complexity. Three.js and other tools depend on unique names, and if users’ workflows depend on names then they should make those names unique. If they’re trying to assign non-unique data like tags, they can use glTF extras just as easily.

> The complexity is O(n) where n is the number of nodes having the same name in a gltf.

I think it’s pretty unlikely that N is ever going to be large with this case, but I’d be open to suggestions on other ways to implement this without naming conflicts.

> if gltf has a lot of nodes with empty name (eg: Voxel model?), user may see serious performance problem.

We’re intentionally not de-duplicating empty names here, so `_1`, `_2`, ... would not be generated.

File: examples/js/loaders/GLTFLoader.js (hunk 1 above)
Comment (donmccurdy) [reply examples/js/loaders/GLTFLoader.js]: Either is fine with me, yes 👍

**Comment by takahirox** [comment]
Just to clarify, I don't block merging. Just my two cents.

> I’d prefer not to preserve the original name in extras

With #16412 we already store original name to `.userData.name`. I meant I suggested glTF exporter (not loader) exports a node with original name if exists.

```javascript
// in GLTFExporter
if (object.userData.name || object.name) gltfNode.name = object.userData.name || object.name;
```

> Three.js and other tools depend on unique names,

Yeah unique names may be less problematic but Three.js and glTF core spec allow duplicate names.

I think better not to lose or change valid information during gltf import-export cycle not to affect user workflow. I know we can't perfectly keep information because Three.js structure doesn't perfectly 1:1 to glTF spec so as much as possible tho.

> if users’ workflows depend on names then they should make those names unique.

The example I said

> For example, if a user has a program out of Three.js processing something only for nodes which have a certain name.

meant there might be program processing multiple nodes which have the same certain name. I know it may be rare but I just wanted to mention that there might be a chance renaming could break user workflow.

> I think it’s pretty unlikely that N is ever going to be large with this case, 

> We’re intentionally not de-duplicating empty names here, so _1, _2, ... would not be generated.

We don't do for empty node names but we do for empty mesh names with `mesh_1, mesh_2, ...`.  In general the number of meshes wouldn't be so large but I guess there might still be a chance. Yeah I know it may be rare so optimizing later may be ok. Adding inline comment about O(n) so far would be good.

**Comment by donmccurdy** [comment]
> I know it may be rare but I just wanted to mention that there might be a chance renaming could break user workflow.

I understand, but it's not just rare – it's not a workflow we've ever seen anyone use before. The glTF spec tries to be clear, but cannot disallow users from doing all possible bad things. That doesn't mean we have to preemptively support bad workflows in three.js, just because they're technically possible. Relying on non-unique names is a bad idea in any three.js-based workflow. I would much rather not add code we have to maintain for this – at least this way it handles things in an easy-to-understand way (by deduping the names) rather than a hard-to-understand way (animation is broken).

> We don't do for empty node names but we do for empty mesh names with mesh_1, mesh_2,...

Hm, I see. So if the mesh had 100 primitives, they'll get the same initial name, and then that's looping over a lot of things to dedupe them. We could fix that case by including an index on each primitive's default name, so they don't conflict?

**Comment by takahirox** [comment]
I was thinking of using `object.uuid` instead of `object.name`. If we do, the loader can be a bit simpler because we can remove unique name creation function and `PropertyBinding.sanitizeNodeName()`. And directly pointing node with uuid seems being more fitting to glTF spec rather than specifying with name. In glTF spec, `channel.target` directly points to a node with node index.

But yes, as you mentioned in the previous thread, if we use uuid animation clip won't work for cloned object because the cloned object will have different uuid. Probably we need to add this type of animation clip cloning utility function. But maybe still user who wants to clone object and reuse animation would be bothered a bit.

```javascript
function cloneAnimationClipForClonedObject(clip, clonedObject) {
  var uuid = clonedObject.uuid;
  var tracks = clip.tracks;
  var newTracks = [];
  for (var i = 0, il = tracks.length; i < il; i++) {
    var track = tracks[i];
    var newName = track.name.replace(/^[^.]*\./, uuid + '.');
    var newTrack = new track.constructor(newName, track.times, track.values, track.getInterpolation());
    newTracks.push(newTrack);
  }
  return new THREE.AnimationClip(clip.name, clip.duration, newTracks);
}
```

So, to be honest I'm still on the fence about renaming. But I can understand your opinion. Renaming is reasonable solution and less problem animation is very useful to users. And I know maybe I'm thinking of very rare (or unlikely happening) use case. So please go ahead with this PR.

> Hm, I see. So if the mesh had 100 primitives, they'll get the same initial name, and then that's looping over a lot of things to dedupe them. We could fix that case by including an index on each primitive's default name, so they don't conflict?

As you know personally I don't really want to aggressively rename but yes I think it'd resolve.

**Comment by takahirox** [comment]
Just in case, let me clarify my opinion. I don't block this change. Yes, we should resolve animation problem in case nodes have non-unique name. Renaming sounds a reasonable solution. Just I wanted to share my two cents.

**Comment by robertlong** [comment]
Right now we're seeing a related bug in Mozilla Spoke. Someone uses an asset from Sketchfab with animations and duplicate names and it breaks.

Example:
https://sketchfab.com/3d-models/christmas-low-poly-scene-7984bd6024db4b89928ddeb625eebac0

We want to be able to export a scene composed of these assets and others using GLTFExporter. I wish we could keep the same name and avoid renaming nodes altogether, that'd probably create less confusion when using the exported asset. If there's any way to make the AnimationMixer not rely on names, or an alternate AnimationMixer that handled references to objects some other way, that'd be ideal, but maybe this solution is good enough.

**Comment by donmccurdy** [comment]
I have no better suggestion than this PR – in my opinion, we should deduplicate node names.

> If there's any way to make the AnimationMixer not rely on names, or an alternate AnimationMixer that handled references to objects some other way, that'd be ideal...

AnimationMixer also supports UUIDs. If you never need to clone an animated object, or are willing to implement custom logic to remap animations to the new UUIDs after cloning, that would work fine for Spoke – just modify GLTFLoader to target animations to a UUID. Unfortunately, this is not a general-purpose solution: most users expect `.clone()` to work with animated objects, and the fact that it does not work naturally is (a) unavoidable as far as I can see, and (b) a more critical problem than appending suffixes to nodes with duplicated names.

**Comment by mrdoob** [comment]
@donmccurdy what are you thought about this PR?

**Comment by mattmattvanvoorst** [comment]
Currently using https://github.com/react-spring/gltfjsx in combination with Fiber to convert a .glb to a React component, but duplicate names in the model are breaking the end result. Would really appreciate this being merged, as I believe it will fix my issue.

It will require some code changes in gltfjsx but I can look at that once this is merged.

**Comment by drcmda** [comment]
> It will require some code changes in gltfjsx but I can look at that once this is merged.

i guess we just need to make gltfloader work under node - which isnt that much work.

@donmccurdy @mrdoob  overall this change would be something i am looking forward to very much. it opens the doors for official, declarative usage of gltf.

something like this is common nowadays in all frameworks

```jsx
const SvgComponent = (props) => (
  <svg width="1em" height="1em" viewBox="0 0 48 1" {...props}>
    <path d="M0 0h48v1H0z" fill="currentColor" fillRule="evenodd" />
  </svg>
)
```

no reason gltfs can't be expressed like that, other than the naming issue:

```jsx
function Model(props) {
  const { nodes, materials } = useLoader(GLTFLoader, '/model.gltf')
  return (
    <group {...props} dispose={null}>
      <group name="Camera" position={[10, 0, 50]} rotation={[Math.PI / 2, 0, 0]}>
        <primitive object={nodes.Camera_Orientation} />
      </group>
      <group name="Sun" position={[100, 50, 100]} rotation={[-Math.PI / 2, 0, 0]}>
        <primitive object={nodes.Sun_Orientation} />
      </group>
      <group name="Cube">
        <mesh material={materials.base} geometry={nodes.Cube_003_0.geometry} />
        <mesh material={materials.inner} geometry={nodes.Cube_003_1.geometry} />
      </group>
    </group>
  )
}
```

**Comment by takahirox** [comment]
I think we should go with this change. We often get bug reports caused by duplicated names from glTF users.

In some use cases, I know some users may not want to rename internally and automatically. But such demand may be relatively more rare case than having animation problems caused by duplicated names. So this change is  would be reasonable solution.

For such users who don't want to rename, there is an option they have their own fork. And I'm thinking of somehow making the use of new plugin(extension) system to provide a way for keeping original name even without forking or complex core (I haven't deeply thought yet tho).

**Comment by mrdoob** [comment]
@donmccurdy could you resolve the conflicts?

**Comment by donmccurdy** [comment]
Rebased ✅

**Comment by takahirox** [comment]
Thanks for rebasing.

E2E check seems to fail on `webgl_instancing_scatter`. The example uses `GLTFLoader`, sorry I haven't looked into the detail yet but can this change affect the example?

**Comment by donmccurdy** [comment]
Thanks! Looks like that failure might be an actual issue, I'll look into it.

**Comment by donmccurdy** [comment]
Ok, fixed. Bit of an edge case to make sure that the parent object gets first shot at the name, rather than its children.

**Review by takahirox — APPROVED** [review]

**Comment by mrdoob** [comment]
Thanks!

**Comment by theRevilot** [comment]
**I was wondering if there is any specific reason to start at index i=1 when renaming nodes within the GLTFLoader, rather than i=0.** The loop I am referring to is located in the assignUniqueName function posted below:

```javascript
GLTFParser.prototype.assignUniqueName = function ( object, originalName ) {

	var name = originalName;

	for ( var i = 1; this.nodeNamesUsed[ name ]; ++ i ) {

		name = originalName + '_' + i;

	}

	this.nodeNamesUsed[ name ] = true;

	object.name = name;

};
```

**I am adressing this because previous ThreeJS versions would start renaming nodes starting at index 0.** As a result the GLTF-Loader caused problems in certain ThreeJS-projects of mine. 
Concerning backwards compability I would therefore suggest to start the loop at index i=0, if there is no specific reason to start at index i=1.

**Comment by drcmda** [comment]
keep in mind though that backward compat is already shot. gltfloader changed two or three times over the last few releases. this hit the react crowd especially since they are not traversing gltf data, they depend on unique names and declarative scenes. if we change this just to fix an index it will not do anything to backward compat since that's already gone out of the window, instead it will just break code again.

**Comment by donmccurdy** [comment]
It's pretty hard to maintain backwards compatibility on generated names, since the order of traversal depends various things. Ideally, it would be best to ensure everything that needs a stable name contains that name in the file. However, the switch from 0- to 1-indexed names probably was not necessary, sorry about that issue! But as @drcmda says, it may just cause more churn to revert it at this point.

**Comment by acmoles** [comment]
I really appreciate all the hard work you're all putting into this... but you broke my workflow ;)

Previously it was possible to load a GLTF file with several skinned meshes and several animation clips and share those clips among them. Upon updating I notice only one mesh is able to use those animations and after rooting around the problem I believe this change is the cause. I haven't figured out why it's causing an issue yet and would appreciate a steer.

[Here ](https://github.com/acmoles/portfolio/blob/master/portfolio/.vuepress/theme/application/three/loadedContent.js#L130) is where I was applying animations to each mesh, published at: https://www.acmoles.com/about/

**Comment by acmoles** [comment]
@mrdoob Here's a fairly minimal fiddle: https://jsfiddle.net/acmoles/oy60a41z/132/

You can change the CDN link from latest to r115 to see what happens - animations working on all models in r115 and stop on latest
(but I think the problem is introduced from r120 - r121, I just couldn't get UNPKG to work for some reason, maybe adblocker)

Thanks in advance for checking it out!

**Comment by FlorentMasson** [comment]
Deduplicating names is a big breaking change for us, I wish there was a way to disable it.
The way described here https://github.com/mrdoob/three.js/issues/15087#issuecomment-751054940 appears to be working. 
Are there any hidden side effects?

**Comment by donmccurdy** [comment]
>  Are there any hidden side effects?

glTF targets animations to specific objects by an internal ID, and does not care what the names of the objects are. three.js targets animation to nodes by name, and assumes that those names are unique within the subtree of the AnimationMixer. So, the hidden side effect of disabling de-duplication of names is that if you are trying to animate anything with a duplicated name, your animation might modify the wrong part of the scene.

The example by @acmoles is basically making use of that mismatch between glTF and three.js, and retargeting animations (which did target a single character in the source glTF file) to different characters, which is possible if you know that the characters have similarly-named joints.

The other workaround we've considered was to set up the animations to target `object.uuid`, which three.js already supports, and then you don't have to de-duplicate names. But then you have another problem — objects cannot be animated after cloning — and the workflow of @acmoles still wouldn't work in this case.

***

I don't have a good solution for this. De-duplicating names ensures that animation plays as expected for people who are just copying animation examples from the three.js tutorials, regardless of the input file. But it does cause problems for those whose workflows depend on objects having specific names, like this case. I don't know how to make both of those "just work", but here are two compromise options:

1. Revert this change, no longer de-duplicating names, and log a warning if any duplicated names are targeted by animation. Then at least the user has a way to figure out why the animation doesn't play, and @acmoles' workflow would be supported, albeit with a noisy warning. Users could either rename things in Blender (lot of work...) or we could provide an opt-in method to do the unique naming.
2. Keep this change, but provide a method to opt out of unique naming.

**Comment by FlorentMasson** [comment]
Thanks for the clarification!

I'm a bit surprised the motivation is fixing duplicate naming in three.js tutorials, wouldn't it best to update them? Ambiguous names can't be a good thing there.

We use mixamo animations which do not have duplicate names, so that's not a concern for us.
But for level design we use Sketchup which generates duplicate names (especially when you use components). We use names to locate specific nodes and we had to use full paths with names from root to leaf - and manually make sure at least the full path is unique. If the de-duplication had been there from the beginning it would certainly have been easier so I can't really speak up against it.
My only worry would be that there's no guarantee a deduplicated node will always have the same name when the node ordering changes. And that, unless I'm mistaken, the uniqueness of names is questionable when names can be empty.

Just a random thought but if duplicated names are only a problem for animation targets, maybe only these names should be deduplicated?

**Comment by drcmda** [comment]
if there's any debate to be had still, i beg you to please not break it again. in my opinion, de-dupe is quite useful and if people can opt out with a plugin that would be fine. we went through a serious of hard breaking changes and finally it has settled for a bit. seeing codebases and hundreds of examples that we made break every release for 2-3 times wasn't fun.

**Comment by donmccurdy** [comment]
> I'm a bit surprised the motivation is fixing duplicate naming in three.js tutorials, wouldn't it best to update them?

There's nothing wrong with the tutorials though — if you load a model and play its animations, users should expect that to just work. But if the model happens to have duplicate names on animated objects, it won't work in three.js without this de-duplication. The same glTF file will work in any other glTF viewer though, it's a perfectly valid file.

I guess I am leaning toward an opt-out, then. Something like:

```js
const loader = new GLTFLoader()
  .setUniqueNames( false )
  .setKTX2Loader( ... )
  .load( ... );
```

**Comment by acmoles** [comment]
The opt-in would be good. Sharing animations among similar skinned meshes is quite a common workflow in game engines, so it's likely to be an expectation for users coming from that world.

**Comment by mattmattvanvoorst** [comment]
I would like to offer the consideration of maybe proving both? A `uuid` and `name` property which are unique, or truthful to the original ID provided in whatever modelling software was used create it?

**Comment by donmccurdy** [comment]
The three.js `object.uuid` is always there in three.js, but has nothing to do with the source glTF model. With the `setUniqueNames(false)` option, any `object.name` value would not be changed from the source glTF model, and should match whatever the source file provided. We might still want to generate names for objects that don't have them in that case, though.

**Pull request merged**